### PR TITLE
PSA MAC computation with _HASH flag implies _MESSAGE.

### LINKED ISSUE #4755: PSA: SIGN_MESSAGE usage does not allow MAC
The key usage flags [`PSA_KEY_USAGE_SIGN_MESSAGE`](https://armmbed.github.io/mbed-crypto/html/api/keys/policy.html#c.PSA_KEY_USAGE_SIGN_MESSAGE) and [`PSA_KEY_USAGE_VERIFY_MESSAGE`](https://armmbed.github.io/mbed-crypto/html/api/keys/policy.html#c.PSA_KEY_USAGE_VERIFY_MESSAGE) allow MAC operations ([`psa_mac_compute`](https://armmbed.github.io/mbed-crypto/html/api/ops/macs.html#c.psa_mac_compute) and [`psa_mac_sign_setup`](https://armmbed.github.io/mbed-crypto/html/api/ops/macs.html#c.psa_mac_sign_setup) in the sign direction, [`psa_mac_verify`](https://armmbed.github.io/mbed-crypto/html/api/ops/macs.html#c.psa_mac_verify) and [`psa_mac_verify_setup`](https://armmbed.github.io/mbed-crypto/html/api/ops/macs.html#c.psa_mac_verify_setup) in the verify direction). The `SIGN_HASH` and `VERIFY_HASH` flags also allow these operations because they imply `SIGN_MESSAGE` and `VERIFY_MESSAGE` respectively.

Looking at the source code, this is not the case: `psa_mac_compute_internal` and `psa_mac_setup` require the `HASH` policies. We forgot this when we marked https://github.com/ARMmbed/mbedtls/issues/3267 as resolved.

Goal of this task: change the requirements. Make sure there are positive test cases for all applicable policies.

### PR DESCRIPTION
**Pull request by mstarzyk-mobica**
Resolves #4755

### CODE REVIEW

File: tests/suites/test_suite_psa_crypto.data
Comment (mstarzyk-mobica) [inline tests/suites/test_suite_psa_crypto.data:1263]: Result from `./test_suite_psa_crypto`
```
PSA MAC sign: RFC4231 Test case 1 - HMAC-SHA-224 .................. PASS
PSA MAC sign: RFC4231 Test case 1 - HMAC-SHA-224 .................. PASS
PSA MAC sign: RFC4231 Test case 1 - HMAC-SHA-224 .................. FAILED
  ( psa_mac_compute( key, alg, input->x, input->len, actual_mac, output_size, &mac_length ) ) == ( expected_status )
  at step 0, line 2088, suites/test_suite_psa_crypto.function
```

File: tests/suites/test_suite_psa_crypto.data
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.data:1253]: Two points:

1. I'm not a fan of the naming: the operation we do is "PSA MAC sign", regardless of which usage bits we set in the key. There's no operation that is conceptually "PSA MAC sign hash" so we shouldn't call it that.
2. I'm not sure it's useful to have three copies of each test. IMO there are two independent things we want to be testing here: (a) whether the result of the operation is correct (using a number of test vectors) and (b) whether permissions on the key are checked properly. And this is already reflected in the tests: with have the test vectors here for (a), and (b) was handled by `mac_key_policy` (modified in the previous commit), so I don't think it needs to be re-done here.

I suggest replacing this commit with one that would keep the test data unchanged, and in the test function just use the `MESSAGE` policy instead of the `HASH` one. We already have test ensuring that `HASH` gets completed to `HASH|MESSAGE` so `MESSAGE` is effectively the minimal policy to use here, so that's we should use when the testing against test vectors.

File: tests/suites/test_suite_psa_crypto.function
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.function:821]: I think this change is correct considering the starting point, but I'm afraid I'm unhappy enough with the starting point that I'm going to ask you to change the logic here: we really should have `expected_status_sign_arg` and `expected_status_verify_arg` as arguments. Then we'd have no `if` statement here and a simple `TEST_EQUAL( status, expected_status_sign );`.

The way it's currently done with the if statement is very misleading. When I look at the data file and see a test case like
```
PSA key policy: MAC, SIGN_MESSAGE
depends_on:PSA_WANT_ALG_HMAC:PSA_WANT_ALG_SHA_256:PSA_WANT_KEY_TYPE_HMAC
mac_key_policy:PSA_KEY_USAGE_SIGN_MESSAGE:PSA_ALG_HMAC(PSA_ALG_SHA_256):PSA_KEY_TYPE_HMAC:"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa":PSA_ALG_HMAC(PSA_ALG_SHA_256):PSA_SUCCESS
```
it is extremely surprising to find out that, before your PR, the test actually asserted that the sign and verify operations would both fail with `PSA_ERROR_NOT_PERMITTED` - the test data clear says `PSA_SUCCESS` and this is absolutely ignored by the test function. After your patch, the test data is exactly the same but now the test asserts that the sign operation returns `PSA_SUCCESS` (as is visible from the data file) and that the verify operation returns `PSA_ERROR_NOT_PERMITTED` (which is completely surprising when looking at the data file which only says `PSA_SUCCESS`).

The test data should end with `PSA_SUCCESS:PSA_ERROR_NOT_PERMITTED` - then it's clear that we expect one operation to succeed and the other to fail.

Sorry, I know it's pre-existing, but I'm afraid I can't approve the PR without fixing this, as the test data is still very misleading.

**Review by mpg — CHANGES_REQUESTED** [review]
I agree that the core fix in the first commit is correct, but I have reservations about the testing strategy: there's a pre-existing design issue with the policy test that IMO needs to be fixed now, and then I don't think we need to re-do policy testing while checking against test vectors.

File: tests/suites/test_suite_psa_crypto.data
Comment (mstarzyk-mobica) [reply tests/suites/test_suite_psa_crypto.data:1253]: re 1: the test name is not really describing the operation, but is rather a short summary of what we are testing. I agree it could be a bit more descriptive, something like `PSA MAC sign, only hash flag`
re 2: the fix from first commit is modifying the functions `psa_mac_compute_internal` and `psa_mac_setup`. Here is a short summary of function calls executed by `test_suite_psa_crypto.function`:
```
mac_sign -> psa_mac_compute -> psa_mac_compute_internal
mac_sign -> psa_mac_sign_setup -> psa_mac_setup
mac_key_policy -> psa_mac_sign_setup -> psa_mac_setup
mac_verify -> psa_mac_verify_setup -> psa_mac_setup
```
`psa_mac_compute` is tested only by the `mac_sign`, so I think that the `mac_sign` tests have to be extended to cover the execution of `psa_mac_compute` with both `_MESSAGE` and `_HASH` flags. This behavior cannot be tested by the `mac_key_policy` alone.

File: tests/suites/test_suite_psa_crypto.function
Comment (mstarzyk-mobica) [reply tests/suites/test_suite_psa_crypto.function]: Changed mac_key_policy function to accept expected status for both sign and verify operations

File: tests/suites/test_suite_psa_crypto.data
Comment (mstarzyk-mobica) [reply tests/suites/test_suite_psa_crypto.data]: Fixed test cases descriptions

Comment (mpg) [reply tests/suites/test_suite_psa_crypto.data]: re 1: thanks for improving the names

re 2: good point, I was assuming `psa_mac_compute()` would be implemented by calling `psa_mac_setup()` at some point, but that's evidently not the case due to drivers, my mistake. I fully agree we want to be testing this function too :)

I'm still not happy with triplicating all the `mac_sign` and `mac_verify` test cases, and still think it would be better to test in an orthogonal way: `mac_key_policy` for all things related to policy, and `mac_sign` and `mac_verify` for functional correctness.

I think that can be achieved by extending the `mac_key_policy` function to also test `psa_mac_compute()` and `psa_mac_verify()`. For `psa_mac_compute()` it's easy: use an all-0 (or any other fixed value) buffer as input, a large enough buffer for the output, and the expected status is the same as for `psa_mac_sign_setup()`. For `psa_mac_verify()` we need a bit more care: re-use the input and output buffers from `psa_mac_compute()`, but for the expected status do something like:
```
if( expected_status_sign != PSA_SUCCESS && expected_status_verify == PSA_SUCCESS )
    TEST_EQUAL( status, PSA_ERROR_INVALID_SIGNATURE );
else
    TEST_EQUAL( status, expected_status_verify );
```

**Review by mpg — CHANGES_REQUESTED** [review]
Thanks for addressing my feedback and pointing out what I missed about testing. I still think the current strategy is not optimal and we could get the same results at a lower cost.

File: tests/suites/test_suite_psa_crypto.data
Comment (mstarzyk-mobica) [reply tests/suites/test_suite_psa_crypto.data]: implemented requested change

**Review by mpg — APPROVED** [review]
Looks good to me! Thanks for bearing with me as I kept requesting structural changes in the tests.

**Comment by mpg** [comment]
Ping @lukgni for second review now that this has one approval.

**Review by lukgni — APPROVED** [review]
LGTM!

**Pull request merged**
